Add go.mod and make gob encodable

costela opened the pull request:
_(copied from #1 because it was mistakenly created from the `master` branch of our fork)_

Hi,

I imagine chances are slim of getting this code merged upstream, but better safe than sorry! :)

Thir small PR does a couple of things:

1. add a `go.mod`.
2. make `Quantile` instances serializable via `encoding/gob`.

The `encoding/gob` support is nice for us because we want to keep stats on long-running p-square estimations across deployemnts, so we need to persist the internal state somehow.

I picked `encoding/gob` (as opposed to e.g. `encoding/json`) because I felt a more opaque representation to be more fitting to internal state that is not supposed to be fiddled with.

WDYT?

costela committed:
add gob encoding/decoding support

costela committed:
go mod init

narqo commented:
Generally, the suggested changes are fine with me, so thank you for the contribution (and I'm sorry, that your previous PR's slipped). 

I wonder if you considered moving all fields from `Quantile` to `internalQuantile`, and make the later the internal data-field of Quantile? I.e. in preudo-Go:

```
type qnt struct {
    P       float64
    Filled  bool
    Pos     [nMarkers]int
}

func (q *qnt) append(v) { ··· }

type Quantile struct {
    data qnt
}

func (q *Quantile) Append(v) {
    if len(q.data.Heights) != nMarkers {
        // ···
    }
    q.data.append(v)
}

func (q *Quantile) GobDecode(v []byte) error {
  return gob.Decode(v, &q.data)
}
```

(to be frank, I'm not sure if it even worth it to put too much thought into it; just wonder about your gut feeling)

costela committed:
refactor: switch to hiding exportable data in sub-struct

This duplicates less code and has the same effect of disallowing
direct access to elements while allowing (un)marshalling.

costela commented:
D'oh :facepalm: That's definitely cleaner! Thanks for the suggestion! (and for considering the PR at all! :sweat_smile:)

PTAL!

narqo reviewed (approved):
Works for me, thank you! 🔥
